sdk/storage/MigrationGuide.md

```diff
@@ -0,0 +1,477 @@
+
+### Client Structure
+
+**The legacy SDK used a stateful model.** There were container and blob objects that held state regarding service resources and required the user to manually call their update methods. But blob contents were not a part of this state and had to be uploaded/downloaded whenever they were to be interacted with. This became increasingly confusing over time, and increasingly susceptible to thread safety issues.
+
+The modern SDK has taken a client-based approach. There are no objects designed to be representations of storage resources, but instead clients that act as your mechanism to interact with your storage resources in the cloud. **Clients hold no state of your resources.** This is most noticable when looking at [blob metadata](#managing-blob-metadata).
+
+The hierarchical structure of Azure Blob Storage can be understood by the following diagram:  
+![Blob Storage Hierarchy](https://docs.microsoft.com/en-us/azure/storage/blobs/media/storage-blobs-introduction/blob1.png)
+
+In the interest of simplifying the API surface, v12 uses three top level clients to match this structure that can be used to interact with a majority of your resources: `BlobServiceClient`, `BlobContainerClient`, and `BlobClient`. Note that blob-type-specific operations can still be accessed by their specific clients, as in v7.5.
+
+#### Migrating from cloud_blob_directory
+
+Note the absence of a v12 equivalent for v7.5's `cloud_blob_directory`. Directories were an SDK-only concept that did not exist in Azure Blob Storage, and which were not brought forwards into the modern Storage SDK. As shown by the diagram in [Client Structure](#client-structure), containers only contain a flat list of blobs, but those blobs can be named and listed in ways that imply a folder-like structure. See our [Listing Blobs in a Container](#listing-blobs-in-a-container) migration samples later in this guide for more information.
+
+For those whose workloads revolve around manipulating directories and heavily relied on the leagacy SDKs abstraction of this structure, consider the [pros and cons of enabling hierarchical namespace](https://docs.microsoft.com/azure/storage/blobs/data-lake-storage-namespace) on your storage account, which would allow switching to the [Data Lake Gen 2 SDK](https://github.com/Azure/azure-sdk-for-cpp/tree/master/sdk/storage/azure-storage-files-datalake), whose migration is not covered in this document.
+
+#### Class Conversion Reference
+
+The following table lists v7.5 classes and their v12 equivalents for quick reference.
+
+| v7.5 | v12 |
+|-------|--------|
+| `cloud_blob_client` | `BlobServiceClient` |
+| `cloud_blob_container`  | `BlobContainerClient` |
+| `cloud_blob_directory` | No equivalent |
+| `cloud_blob` | `BlobClient` |
+| `cloud_block_blob` | `BlockBlobClient` |
+| `cloud_page_blob` | `PageBlobClient` |
+| `cloud_append_blob` | `AppendBlobClient` |
+
+## Migration Samples
```

Comment:
Lease is also different in usage. Might need a section on that.
LRO (poller for copy) is another one in my mind.

Reply:
Track1 SDK doesn't even support waiting on LRO. How can anybody possibly migrate that functionality?

Reply:
Migration from customer code that sleeps and waits until copy status returns finished to this more elegant approach of PollUntilDone is somehow worth mentioning to me. But feel free to disagree and ignore.